pkg/mysql/mysql_test.go

```diff
@@ -0,0 +1,352 @@
+						ObjectMeta: metav1.ObjectMeta{
+							Annotations: expectedAnnotations,
+						},
+						Spec: corev1.PodSpec{
+							InitContainers: []corev1.Container{
+								{
+									Image: initImage,
+								},
+							},
+							TerminationGracePeriodSeconds: pointerInt64(30),
+							Volumes: append(expectedVolumes(),
+								corev1.Volume{
+									Name: "datadir",
+									VolumeSource: corev1.VolumeSource{
+										HostPath: &corev1.HostPathVolumeSource{},
+									},
+								},
+							),
+						},
+					},
+					ServiceName: "cluster1-mysql",
+				},
+			},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			cr := &apiv1alpha1.PerconaServerMySQL{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cluster1",
+					Namespace: "test-ns",
+				},
+				Spec: apiv1alpha1.PerconaServerMySQLSpec{
+					MySQL: tt.mysqlSpec,
+				},
+			}
+
+			sts := StatefulSet(cr, initImage, configHash, tlsHash, nil)
+
+			assert.NotNil(t, sts)
+			assert.Equal(t, tt.expectedStatefulSet.Name, sts.Name)
+			assert.Equal(t, tt.expectedStatefulSet.Namespace, sts.Namespace)
+			assert.Equal(t, tt.expectedStatefulSet.Spec.Replicas, sts.Spec.Replicas)
+
+			assert.Equal(t, tt.expectedStatefulSet.Spec.Template.Annotations, sts.Spec.Template.Annotations)
+
+			assert.Equal(t, tt.expectedStatefulSet.Spec.ServiceName, sts.Spec.ServiceName)
+
+			assert.Equal(t, tt.expectedStatefulSet.Spec.Template.Spec.Volumes, sts.Spec.Template.Spec.Volumes)
+
+			initContainers := sts.Spec.Template.Spec.InitContainers
+			assert.Len(t, initContainers, 1)
+			assert.Equal(t, initImage, initContainers[0].Image)
+
+			assert.Equal(t, tt.expectedStatefulSet.Spec.Template.Spec.TerminationGracePeriodSeconds, sts.Spec.Template.Spec.TerminationGracePeriodSeconds)
+
+			assert.Equal(t, tt.expectedStatefulSet.Spec.VolumeClaimTemplates, sts.Spec.VolumeClaimTemplates)
+		})
+	}
+}
+
+func expectedVolumes() []corev1.Volume {
+	return []corev1.Volume{
+		{
+			Name: "bin",
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		},
+		{
+			Name: "mysqlsh",
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		},
+		{
+			Name: "users",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: "internal-cluster1",
+				},
+			},
+		},
+		{
+			Name: "tls",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: "",
+				},
+			},
+		},
+		{
+			Name: "config",
+			VolumeSource: corev1.VolumeSource{
+				Projected: &corev1.ProjectedVolumeSource{
+					Sources: []corev1.VolumeProjection{
+						{
+							ConfigMap: &corev1.ConfigMapProjection{
+								LocalObjectReference: corev1.LocalObjectReference{
+									Name: "cluster1-mysql",
+								},
+								Items: []corev1.KeyToPath{
+									{
+										Key:  "my.cnf",
+										Path: "my-config.cnf",
+									},
+								},
```

Comment:
we can use `ptr.To` from "k8s.io/utils/ptr"

Reply:
I think we can avoid importing a whole library just for converting to pointers. Checked their implementation and I think we can improve our own however by using generics as well.

```
func toPointer[T any](v T) *T {
	return &v
}
```
WDYT about this: https://github.com/percona/percona-server-mysql-operator/pull/933/commits/b40c7bb707779c98032a7e7e3ae149ed29d93cb6?

Reply:
discussed with @egegunes, since the library is already imported indirectly from our dependencies maybe it is ok to use it. https://github.com/percona/percona-server-mysql-operator/pull/933/commits/998557120465f1a52aa84deafe64c41ea84e9e7e